core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -594,7 +635,7 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
                     txnMetadata.inLock {
                       if (txnMetadata.producerId != producerId)
                         Left(Errors.INVALID_PRODUCER_ID_MAPPING)
-                      else if (txnMetadata.producerEpoch != producerEpoch)
+                      else if (txnMetadata.producerEpoch != producerEpoch && (requestIsAtLeastTransactionsV2 && txnMetadata.producerEpoch != producerEpoch + 1))
```

Comment:
Should it be `txnMetadata.producerEpoch != producerEpoch && (!requestIsAtLeastTransactionsV2 || txnMetadata.producerEpoch != producerEpoch + 1)`?  As written we effectively ignore epoch check when `requestIsAtLeastTransactionsV2==false`.

Reply:
I think I got my logic backwards here. I will take another look since as is the parentheses don't make sense.